src/test/ui/array-slice-vec/subslice-patterns-const-eval-match.rs

```diff
@@ -0,0 +1,110 @@
+// This test comprehensively checks the passing static and dynamic semantics
+// of subslice patterns `..`, `x @ ..`, `ref x @ ..`, and `ref mut @ ..`
+// in slice patterns `[$($pat), $(,)?]` .
+
+// run-pass
+
+#![feature(slice_patterns, const_fn, const_if_match)]
+#[derive(PartialEq, Debug, Clone)]
+struct N(u8);
+
+#[derive(PartialEq, Debug, Clone)]
+struct Z;
+
+macro_rules! n {
+    ($($e:expr),* $(,)?) => {
+        [$(N($e)),*]
+    }
+}
+
+macro_rules! zed {
+    ($e:expr) => { Z }
+}
+
+macro_rules! z {
```

Comment:
The test this is derived from is so dense and lacking of comments, but in the spirit of not repeating my mistakes.. it would be useful to have a few comments (e.g. here because `zed` doesn't use `$e`, which is interesting).